### PR TITLE
trying out a component base idea

### PR DESCRIPTION
**Pull request by flynnplatt**
DO NOT MERGE

Place to discuss and test component inheritance strategies

### CODE REVIEW

**Comment by AaronPlave** [comment]
I'm not entirely sure if having this base container strategy buys us anything we don't already have since right now you can just extend the existing component and override the render function. Or am I missing something? Also one potential common use case to think of is adding a button to the layer controls component – what's the best way of doing that from a cmc app dev perspective?

**Comment by flynnplatt** [comment]
this particular test is more the idea of formalizing our approach to component re-use. Like the LayerControlBase wouldn't even work on its own because it doesn't connect to the redux state. It's basically us saying, "This component should do X,Y, and Z but we don't care what it looks like".

As to adding a component, this approach _could_ help with that as it would create two files, one for all of the things a component will do, and another for how the component will look. But that may also end up being messy. Plus I don't know where else this sort of issue comes up in CMC

**Comment by AaronPlave** [comment]
Gotcha. I think the layer menu is probably our most complex and layered (hah) component, not sure if many other components really run into this issue much. One other idea I had was breaking the render functions up (taking adding a button to layer control container as ex) so that you wouldn't have to copy paste a huge render function just to swap out the core layer control container for your own.

**Comment by flynnplatt** [comment]
What about extending that idea with the multiple render functions with actual new components? Like,

```JSX
import {LayerButtons} from "_core/components/LayerMenu";
export class LayerControl extends LayerControlBase {
  renderButtons(){
    return (<LayerButtons />);
  }
  render() {
    return (
      ...
      {this.renderButtons()}
      ...
    );
  }
}
```

That way you have the flexibility of replacing individual pieces of the rendered component and we still have isolated component pieces

File: src/_core/components/LayerMenu/LayerControlBase.js [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+LayerControlBase.propTypes = {
+    mapActions: PropTypes.object.isRequired,
+    layer: PropTypes.object.isRequired,
+    activeNum: PropTypes.number.isRequired,
+    palette: PropTypes.object
+};
+
+LayerControlBase.mapDispatchToProps = function(dispatch) {
+    return {
+        mapActions: bindActionCreators(mapActions, dispatch)
```
Comment (AaronPlave) [inline src/_core/components/LayerMenu/LayerControlBase.js]: If we do split out the base component here do we want to pass in mapActions as a prop?

**Comment by AaronPlave** [comment]
Yup, that's what I was thinking. I could go either way on the base component idea.

File: src/_core/components/LayerMenu/LayerControlBase.js (hunk 1 above)
Comment (flynnplatt) [reply src/_core/components/LayerMenu/LayerControlBase.js]: I dunno. I like the idea of not having to deal with any part of state/dispatch mapping if you inherit from the base. But then we'd need a way to easily override actions downstream

**Comment by flynnplatt** [comment]
closing this in favor of #69 #68 and other more concrete work

**Pull request closed without merging**